File: backend/services/trade_adapters/ethereal.py
"""Ethereal DEX trade adapter — EIP-712 linked signer.

Users create a "Linked Signer" (any ETH keypair) and link it to their
subaccount via EIP-712 signature. The signer can trade but cannot withdraw.

creds = {"address": "0x... (subaccount)", "private_key": "0x... (linked signer key)"}
"""
from __future__ import annotations

import logging
import time
from typing import Any

import httpx

logger = logging.getLogger("avalant.trade.ethereal")

BASE = "https://api.ethereal.trade"


class EtherealAdapter:

    @staticmethod
    def _symbol(s: str) -> str:
        return s.upper()

    @classmethod
    async def _signed_request(cls, creds: dict, method: str, path: str,
                               params: dict | None = None, body: dict | None = None) -> Any:
        """EIP-712 signed request via linked signer."""
        try:
            from eth_account import Account
        except ImportError:
            raise RuntimeError("eth_account package required for Ethereal trading")

        private_key = creds.get("private_key") or creds.get("api_secret") or ""
        if not private_key:
            raise RuntimeError("Ethereal requires a linked signer private key")

        acct = Account.from_key(private_key)
        ts = str(int(time.time() * 1e9))  # nanoseconds

        # Simplified auth: header-based signature
        import json as _json
        sign_payload = f"{method}{path}{ts}{_json.dumps(body or {}, separators=(',', ':'))}"
        from eth_account.messages import encode_defunct
        signed = acct.sign_message(encode_defunct(text=sign_payload))

        headers = {
            "Content-Type": "application/json",
            "X-Ethereal-Address": creds.get("address") or creds.get("api_key") or "",
            "X-Ethereal-Timestamp": ts,
            "X-Ethereal-Signature": signed.signature.hex(),
        }

        from backend.services.trade_adapters._http import http_client
        client = http_client(BASE, timeout=10.0)
        if method == "GET":
            r = await client.get(path, params=params, headers=headers)
        elif method == "POST":
            r = await client.post(path, json=body, headers=headers)
        else:
            raise ValueError(method)
        if r.status_code >= 400:
            raise RuntimeError(f"Ethereal {r.status_code}: {r.text[:200]}")
        return r.json()

    @classmethod
    async def fetch_balance(cls, creds: dict) -> dict:
        address = creds.get("address") or creds.get("api_key") or ""
        async with httpx.AsyncClient(timeout=10) as c:
            # Two-step lookup matches the working portfolio provider:
            #   1) /v1/subaccount?sender=ADDR → { "data": [{id, ...}] }
            #   2) /v1/subaccount/balance?subaccountId=ID → { "data": [{tokenName, available, totalUsed, ...}] }
            r = await c.get(f"{BASE}/v1/subaccount", params={"sender": address})
            if r.status_code >= 400:
                raise RuntimeError(f"Ethereal {r.status_code}: {r.text[:200]}")
            subs = (r.json() or {}).get("data") or []
            if not subs:
                return {"usdt": 0.0, "spot_usd": 0.0, "futures_usd": 0.0}
            sub_id = subs[0].get("id")
            if not sub_id:
                return {"usdt": 0.0, "spot_usd": 0.0, "futures_usd": 0.0}
            r2 = await c.get(f"{BASE}/v1/subaccount/balance", params={"subaccountId": sub_id})
            if r2.status_code >= 400:
                raise RuntimeError(f"Ethereal balance {r2.status_code}: {r2.text[:200]}")
            items = (r2.json() or {}).get("data") or []
            total = 0.0
            for it in items:
                # Ethereal labels their stable as "USD" (not USDT/USDC).
                # Accept all common stable token names.
                token = (it.get("tokenName") or "").upper()
                if token not in ("USDT", "USDC", "USD", "USDE", "BUSD"):
                    continue
                try:
                    avail = float(it.get("available") or 0)
                    used = float(it.get("totalUsed") or 0)
                    amount = float(it.get("amount") or 0)
                    total += (avail + used) or amount
                except (TypeError, ValueError):
                    continue
            # Ethereal is futures-only.
            return {"usdt": total, "spot_usd": 0.0, "futures_usd": total}

    @classmethod
    async def validate_key(cls, creds: dict, need_trade: bool = False) -> dict:
        out = {"can_read": False, "can_trade": False, "balance_usdt": None, "error": None}
        try:
            bal = await cls.fetch_balance(creds)
            out["can_read"] = True
            out["balance_usdt"] = bal.get("usdt", 0)
        except Exception as e:
            out["error"] = f"Ethereal: {str(e)[:180]}"
            return out
        if need_trade:
            pk = creds.get("private_key") or creds.get("api_secret")
            if not pk:
                out["error"] = "Linked signer private key required for trading"
            else:
                out["can_trade"] = True
        return out

    @classmethod
    async def set_leverage(cls, creds: dict, symbol: str, leverage: int, margin_mode: str) -> None:
        pass  # Ethereal manages leverage per-position at order time

    @classmethod
    async def place_order(cls, creds: dict, symbol: str, side: str, quantity: float,
                          leverage: int = 1, margin_mode: str = "isolated") -> dict:
        body = {
            "symbol": cls._symbol(symbol),
            "side": side,
            "type": "market",
            "quantity": str(quantity),
        }
        r = await cls._signed_request(creds, "POST", "/v1/order", body=body)
        return {"order_id": str(r.get("orderId", r.get("id", ""))), "avg_price": 0.0}

    @classmethod
    async def close_position(cls, creds: dict, symbol: str, side: str) -> dict:
        """Reduce-only market order. Previously called place_order without
        a reduce flag, which would open an opposing position in hedge mode."""
        positions = await cls.list_positions(creds, symbol)
        if not positions:
            return {"order_id": None, "closed_qty": 0, "realized_pnl_usd": 0}
        p = positions[0]
        close_side = "sell" if p["side"] == "buy" else "buy"
        body = {
            "symbol": cls._symbol(symbol),
            "side": close_side,
            "type": "market",
            "quantity": str(p["quantity"]),
            "reduceOnly": True,
        }
        r = await cls._signed_request(creds, "POST", "/v1/order", body=body)
        return {
            "order_id": str(r.get("orderId", r.get("id", ""))),
            "closed_qty": p["quantity"],
            "realized_pnl_usd": p.get("unrealized_pnl_usd", 0),
        }

    @classmethod
    async def list_positions(cls, creds: dict, symbol: str | None = None) -> list[dict]:
        address = creds.get("address") or creds.get("api_key") or ""
        async with httpx.AsyncClient(timeout=10) as c:
            r = await c.get(f"{BASE}/v1/subaccount", params={"sender": address})
            if r.status_code >= 400:
                return []
            j = r.json()
            # Handle {"subaccounts":[{positions}]} shape too
            if isinstance(j, dict) and j.get("subaccounts"):
                j = j["subaccounts"][0] if j["subaccounts"] else {}
        out = []
        for p in j.get("positions", []):
            qty = float(p.get("size", 0) or 0)
            if qty == 0:
                continue
            sym = str(p.get("symbol", p.get("productSymbol", "")))
            if symbol and sym.upper() != symbol.upper():
                continue
            out.append({
                "exchange": "ethereal",
                "symbol": sym,
                "side": "buy" if qty > 0 else "sell",
                "quantity": abs(qty),
                "entry_price": float(p.get("entryPrice", 0) or 0),
                "mark_price": float(p.get("markPrice", 0) or 0),
                "unrealized_pnl_usd": float(p.get("unrealizedPnl", 0) or 0),
                "leverage": 1,
                "position_id": sym,
            })
        return out

    @classmethod
    async def get_public_max_leverage(cls, symbol: str) -> int:
        return 20

    @classmethod
    async def preflight(cls, creds: dict, symbol: str, quantity: float, leverage: int) -> dict:
        return {"ok": True, "qty_rounded": quantity}

File: go-fetcher/internal/trade/ethereal/ethereal.go
// Ethereal DEX trade adapter — linked-signer auth via personal_sign.
//
// Port of `backend/services/trade_adapters/ethereal.py`.
//
// Auth: a "Linked Signer" (any ETH keypair) that the user has bound to
// their subaccount on ethereal.trade. The signer can trade but cannot
// withdraw. We do NOT submit EIP-712 typed data — Python's adapter
// uses the simpler personal_sign over a flat string:
//
//	sign_payload = METHOD || PATH || TIMESTAMP_NS || JSON(body)
//	sig          = personal_sign(sign_payload, linkedSignerKey)
//
// Headers
//
//	X-Ethereal-Address:   <subaccount addr>
//	X-Ethereal-Timestamp: <ns>
//	X-Ethereal-Signature: <0x-prefixed 65-byte hex>
//
// Quirks
//
//   - Timestamp is in NANOseconds.
//   - GET /v1/subaccount?address=… is unsigned (public read).
//   - Public WS isn't usable; the adapter drives only trade actions.
package ethereal

import (
	"bytes"
	"context"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/ethereum/go-ethereum/crypto"

	"github.com/bashyrov/wallet-monitor/go-fetcher/internal/trade"
)

const baseURL = "https://api.ethereal.trade"

type Adapter struct {
	httpClient *http.Client
}

func New() *Adapter {
	return &Adapter{
		httpClient: &http.Client{
			Timeout: 15 * time.Second,
			Transport: &http.Transport{
				MaxIdleConnsPerHost: 8,
				IdleConnTimeout:     60 * time.Second,
			},
		},
	}
}

func init() { trade.Register("ethereal", New()) }

func (a *Adapter) Name() string { return "ethereal" }

// ── Signing ──────────────────────────────────────────────────────────────

// signPersonal returns the 0x-prefixed personal_sign hex for the given
// payload. Mirrors `eth_account.messages.encode_defunct(text=...)`
// followed by `Account.sign_message`.
func signPersonal(payload, privKeyHex string) (string, error) {
	priv, err := crypto.HexToECDSA(strings.TrimPrefix(privKeyHex, "0x"))
	if err != nil {
		return "", fmt.Errorf("parse private key: %w", err)
	}
	prefix := []byte(fmt.Sprintf("\x19Ethereum Signed Message:\n%d", len(payload)))
	msg := append(prefix, []byte(payload)...)
	digest := crypto.Keccak256(msg)
	sig, err := crypto.Sign(digest, priv)
	if err != nil {
		return "", fmt.Errorf("sign: %w", err)
	}
	if len(sig) != 65 {
		return "", fmt.Errorf("unexpected sig length %d", len(sig))
	}
	sig[64] += 27
	return "0x" + hex.EncodeToString(sig), nil
}

// signedRequest authenticates with the linked-signer key. GETs that
// don't take auth (public reads) call doRequest directly.
func (a *Adapter) signedRequest(
	ctx context.Context, creds trade.Creds, method, path string,
	body map[string]any,
) (json.RawMessage, error) {
	if creds.APISecret == "" {
		return nil, errUser("linked signer private key required")
	}
	tsNs := strconv.FormatInt(time.Now().UnixNano(), 10)

	bodyBytes := []byte("{}")
	if body != nil {
		b, err := json.Marshal(body)
		if err != nil {
			return nil, errInternal("marshal body", err)
		}
		bodyBytes = b
	}
	payload := method + path + tsNs + string(bodyBytes)
	sig, err := signPersonal(payload, creds.APISecret)
	if err != nil {
		return nil, errInternal("personal sign", err)
	}

	address := creds.APIKey
	if address == "" {
		address = creds.Wallet
	}

	var bodyReader io.Reader
	if body != nil {
		bodyReader = bytes.NewReader(bodyBytes)
	}
	req, err := http.NewRequestWithContext(ctx, method, baseURL+path, bodyReader)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Ethereal-Address", address)
	req.Header.Set("X-Ethereal-Timestamp", tsNs)
	req.Header.Set("X-Ethereal-Signature", sig)

	resp, err := a.httpClient.Do(req)
	if err != nil {
		return nil, &trade.Error{Kind: trade.KindTransient, Message: err.Error(), Cause: err}
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(resp.Body)
	if resp.StatusCode >= 400 {
		return nil, parseError(resp.StatusCode, raw)
	}
	if len(raw) == 0 {
		return json.RawMessage("{}"), nil
	}
	return raw, nil
}

func (a *Adapter) publicGet(ctx context.Context, path string, qs url.Values) (json.RawMessage, error) {
	u := baseURL + path
	if len(qs) > 0 {
		u += "?" + qs.Encode()
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u, nil)
	if err != nil {
		return nil, err
	}
	resp, err := a.httpClient.Do(req)
	if err != nil {
		return nil, &trade.Error{Kind: trade.KindTransient, Message: err.Error(), Cause: err}
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(resp.Body)
	if resp.StatusCode >= 400 {
		return nil, parseError(resp.StatusCode, raw)
	}
	return raw, nil
}

func parseError(status int, body []byte) *trade.Error {
	var env struct {
		Error   string `json:"error"`
		Message string `json:"message"`
	}
	_ = json.Unmarshal(body, &env)
	msg := env.Message
	if msg == "" {
		msg = env.Error
	}
	if msg == "" {
		msg = strings.TrimSpace(string(body))
	}
	if status == 429 {
		return &trade.Error{Kind: trade.KindRateLimit, Message: msg}
	}
	return &trade.Error{Kind: trade.KindExchange, Message: msg}
}

// ── Adapter methods ──────────────────────────────────────────────────────

func (a *Adapter) GetBalance(ctx context.Context, creds trade.Creds) (*trade.Balance, error) {
	address := creds.APIKey
	if address == "" {
		address = creds.Wallet
	}
	if address == "" {
		return nil, errUser("subaccount address required")
	}
	body, err := a.publicGet(ctx, "/v1/subaccount", url.Values{"address": []string{address}})
	if err != nil {
		return nil, err
	}
	var resp struct {
		Equity       json.Number `json:"equity"`
		AccountValue json.Number `json:"accountValue"`
	}
	if err := json.Unmarshal(body, &resp); err != nil {
		return nil, errInternal("parse subaccount", err)
	}
	v, _ := resp.Equity.Float64()
	if v == 0 {
		v, _ = resp.AccountValue.Float64()
	}
	return &trade.Balance{TotalUSD: v, AvailableUSD: v}, nil
}

func (a *Adapter) SetLeverage(_ context.Context, _ trade.Creds, _ trade.LeverageRequest) error {
	return nil // Ethereal sets leverage at order time, not as a separate call.
}

func (a *Adapter) PlaceOrder(ctx context.Context, creds trade.Creds, req trade.OpenRequest) (*trade.Result, error) {
	if err := req.Validate(); err != nil {
		return nil, err
	}
	body, err := a.signedRequest(ctx, creds, http.MethodPost, "/v1/order", map[string]any{
		"symbol":   strings.ToUpper(req.Symbol),
		"side":     string(req.Side),
		"type":     "market",
		"quantity": qtyString(req.Quantity),
	})
	if err != nil {
		return nil, err
	}
	id := extractOrderID(body)
	return &trade.Result{
		OrderID:   id,
		Symbol:    req.Symbol,
		Side:      req.Side,
		Quantity:  req.Quantity,
		Status:    "NEW",
		CreatedAt: time.Now().UTC(),
		Raw:       body,
	}, nil
}

func (a *Adapter) ClosePosition(ctx context.Context, creds trade.Creds, req trade.CloseRequest) (*trade.Result, error) {
	if req.Symbol == "" {
		return nil, errUser("symbol required")
	}
	positions, err := a.ListPositions(ctx, creds, req.Symbol)
	if err != nil {
		return nil, err
	}
	if len(positions) == 0 {
		return &trade.Result{Symbol: req.Symbol, Status: "FLAT"}, nil
	}
	p := positions[0]
	closeSide := trade.SideSell
	if p.Side == trade.SideSell {
		closeSide = trade.SideBuy
	}
	body, err := a.signedRequest(ctx, creds, http.MethodPost, "/v1/order", map[string]any{
		"symbol":     strings.ToUpper(req.Symbol),
		"side":       string(closeSide),
		"type":       "market",
		"quantity":   qtyString(p.Quantity),
		"reduceOnly": true,
	})
	if err != nil {
		return nil, err
	}
	id := extractOrderID(body)
	return &trade.Result{
		OrderID:   id,
		Symbol:    req.Symbol,
		Side:      closeSide,
		Quantity:  p.Quantity,
		Status:    "NEW",
		CreatedAt: time.Now().UTC(),
		Raw:       body,
	}, nil
}

func (a *Adapter) ListPositions(ctx context.Context, creds trade.Creds, symbol string) ([]trade.Position, error) {
	address := creds.APIKey
	if address == "" {
		address = creds.Wallet
	}
	if address == "" {
		return nil, errUser("subaccount address required")
	}
	body, err := a.publicGet(ctx, "/v1/subaccount", url.Values{"address": []string{address}})
	if err != nil {
		return nil, err
	}
	var resp struct {
		Positions []struct {
			Symbol         string      `json:"symbol"`
			ProductSymbol  string      `json:"productSymbol"`
			Size           json.Number `json:"size"`
			EntryPrice     json.Number `json:"entryPrice"`
			MarkPrice      json.Number `json:"markPrice"`
			UnrealizedPnl  json.Number `json:"unrealizedPnl"`
		} `json:"positions"`
	}
	if err := json.Unmarshal(body, &resp); err != nil {
		return nil, errInternal("parse positions", err)
	}
	want := strings.ToUpper(symbol)
	out := make([]trade.Position, 0, len(resp.Positions))
	for _, p := range resp.Positions {
		sz, _ := p.Size.Float64()
		if sz == 0 {
			continue
		}
		sym := p.Symbol
		if sym == "" {
			sym = p.ProductSymbol
		}
		if want != "" && strings.ToUpper(sym) != want {
			continue
		}
		side := trade.SideBuy
		if sz < 0 {
			side = trade.SideSell
		}
		entry, _ := p.EntryPrice.Float64()
		mark, _ := p.MarkPrice.Float64()
		upnl, _ := p.UnrealizedPnl.Float64()
		out = append(out, trade.Position{
			Symbol:        sym,
			Side:          side,
			Quantity:      abs(sz),
			EntryPrice:    entry,
			MarkPrice:     mark,
			UnrealizedPnL: upnl,
			Leverage:      1,
		})
	}
	return out, nil
}

// ── Helpers ──────────────────────────────────────────────────────────────

// extractOrderID handles both `{"orderId": "..."}` and `{"id": "..."}`
// (Python adapter falls back to `id` if `orderId` is absent).
func extractOrderID(body []byte) string {
	var resp struct {
		OrderID json.RawMessage `json:"orderId"`
		ID      json.RawMessage `json:"id"`
	}
	if err := json.Unmarshal(body, &resp); err != nil {
		return ""
	}
	if id := unquoteOrNumber(resp.OrderID); id != "" {
		return id
	}
	return unquoteOrNumber(resp.ID)
}

func unquoteOrNumber(raw json.RawMessage) string {
	s := strings.TrimSpace(string(raw))
	if s == "" || s == "null" {
		return ""
	}
	if strings.HasPrefix(s, `"`) && strings.HasSuffix(s, `"`) && len(s) >= 2 {
		return s[1 : len(s)-1]
	}
	return s
}

func qtyString(q float64) string {
	s := strconv.FormatFloat(q, 'f', 8, 64)
	if strings.Contains(s, ".") {
		s = strings.TrimRight(s, "0")
		s = strings.TrimRight(s, ".")
		if s == "" {
			s = "0"
		}
	}
	return s
}

func abs(f float64) float64 {
	if f < 0 {
		return -f
	}
	return f
}

func errUser(msg string, args ...any) *trade.Error {
	return &trade.Error{Kind: trade.KindUser, Message: fmt.Sprintf(msg, args...)}
}

func errInternal(msg string, cause error) *trade.Error {
	return &trade.Error{Kind: trade.KindInternal, Message: msg, Cause: cause}
}

var _ trade.Adapter = (*Adapter)(nil)
